src/mem3/src/mem3_db_doc_updater.erl

```diff
@@ -0,0 +1,107 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(mem3_db_doc_updater).
+
+-behaviour(gen_server).
+
+-export([
+    get_db_doc/1,
+    update_db_doc/1,
+
+    start_link/0,
+
+    init/1,
+    handle_call/3,
+    handle_cast/2
+]).
+
+-include_lib("couch/include/couch_db.hrl").
+
+% Early return shortcut
+%
+-define(THROW(RES), throw({reply, RES, nil})).
+
+get_db_doc(DocId) when is_binary(DocId) ->
+    Timeout = shard_update_timeout_msec(),
+    gen_server:call(first_node(), {get_db_doc, DocId}, Timeout).
+
+update_db_doc(#doc{} = Doc) ->
+    Timeout = shard_update_timeout_msec(),
+    gen_server:call(first_node(), {update_db_doc, Doc}, Timeout).
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init(_) ->
+    {ok, nil}.
+
+handle_call({get_db_doc, DocId}, _From, nil = St) ->
+    {reply, get_db_doc_int(DocId), St};
+handle_call({update_db_doc, #doc{} = Doc}, _From, nil = St) ->
+    {reply, update_db_doc_int(Doc), St};
+handle_call(Msg, _From, nil = St) ->
+    {stop, {invalid_call, Msg}, invalid_call, St}.
+
+handle_cast(Msg, nil = St) ->
+    {stop, {invalid_cast, Msg}, St}.
+
+% Private
+
+update_db_doc_int(#doc{} = Doc) ->
+    ok = validate_coordinator(),
+    couch_util:with_db(mem3_sync:shards_db(), fun(Db) ->
+        try
+            Res = couch_db:update_doc(Db, Doc, [?ADMIN_CTX]),
+            ok = replicate_to_all_nodes(shard_update_timeout_msec()),
+            Res
+        catch
+            conflict ->
+                ?THROW({error, conflict})
```

Comment:
throwing instead of returning a reply tuple in the handle_call callback? does that work?

Reply:
At least they documented it so it’s official. It’s been a hidden feature for years